app/oidc/oidc.py

```diff
@@ -0,0 +1,66 @@
+import functools
+from abc import ABC, abstractmethod
+from datetime import datetime, timedelta, timezone
+from typing import Callable, ParamSpec
+
+import pytz
+from google.auth.credentials import Credentials
+from google.auth.transport.requests import Request
+from structlog import get_logger
+
+from app.settings import OIDC_TOKEN_LEEWAY_IN_SECONDS
+
+P = ParamSpec("P")
+
+logger = get_logger()
+TTL = 3600 - OIDC_TOKEN_LEEWAY_IN_SECONDS  # 1 hour minus leeway in seconds
+
+
+def get_expiry_from_ttl(ttl: int) -> datetime:
+    return datetime.now(timezone.utc) - timedelta(seconds=ttl)
+
+
+def is_naive(dt: datetime) -> bool:
+    return dt.tzinfo is None or dt.tzinfo.utcoffset(dt) is None
+
+
+def refresh_oidc_credentials(
+    func: Callable[P, Credentials]
+) -> Callable[P, Credentials]:
+    """
+    This decorator is used on implementations of the abstract method OIDCCredentialsService.get_credentials.
+    This decorated method uses a ttl_cache decorator to discard fetched credentials after a given time.
+    However, the TTL of the credentials should be determined by inspecting the expiry of the credentials
+     themselves, not relying on the ttl_cache.
+    As such, this decorator ensures the credentials returned from the ttl_cache are valid and refreshed
+     via inspecting the expiry property of the credentials.
+    Overall this ensures the caller is always returned valid credentials and credentials are cached to minimise
+     API calls for credentials as well as discarded once invalid.
+    """
+
+    @functools.wraps(func)
+    def wrapper_refresh_oidc_credentials(
+        *args: P.args, **kwargs: P.kwargs
+    ) -> Credentials:
+        credentials = func(*args, **kwargs)
+        expiry = (
+            pytz.utc.localize(dt=credentials.expiry)
+            if is_naive(credentials.expiry)
+            else credentials.expiry
+        )
+        if expiry < get_expiry_from_ttl(TTL):
+            logger.info("refreshing oidc credentials")
+            credentials.refresh(Request())
```

Comment:
Given this refresh is called outside the function that has the cache wrapper, it isn't clear how exactly cache is working given we are not calling the function.

We should document that this is working because `credentials` references the same cached object in memory, therefore when refreshing this credentials object, it is actually the same object the cache holds. We should also document this must be used with in-memory cache as things like redis wouldn't work with this pattern.

I do like the decorator pattern but I'm not sure I'm a fan of the cache working because it happens to update the same object in memory. If no one else has a concerned with this, I'm happy to yield (just make sure we have through tests using the `cache_info()` method and good documentation).

Reply:
Yeah the existence of the TTL cache ensures that if a token goes unused for over the credential's TTL, it is discarded behind the scenes. 
The reason for the refresh when requesting credentials, ensures that even if a returned cached result has 0.1 seconds of lifetime left, the requesting function has credentials which are guaranteed to be valid for the proceeding usage.  
It also assumes the credentials acquired from the decorated function have this refresh method, so should only decorate functions which return google OIDC credentials in whatever implementation that may be.

Reply:
The usage of why we refresh was clear, but how the cache was being updated with the refreshed credentials isn't. If it has expired then refreshing is correct.

My main point being:

> We should document that this is working because credentials references the same cached object in memory, therefore when refreshing this credentials object, it is actually the same object the cache holds. We should also document this must be used with in-memory cache as things like redis wouldn't work with this pattern. 

The fine detail of the cache working before it is referring the same object in memory is the gotcha. We need to make this clear through some comments.

Reply:
For context for others, the current pattern doesn't quite work as expected. Since the credentials are refreshed outside the function, the ttl of the cache is never extended and we will end up referring the cache at 55 mins, then again at 60 mins enough though the refresh at 55 minutes, should have extended it for another 55 minutes.

Details:
- At 55 minutes, the cache is classed as expired based on our check, so the credential gets refreshed without calling the function. So credential is not valid for another hours (55 mins based on our check)
- A request after would call the function and generate the credential again because the TTL cache has expired even though we refreshed the credentials moments ago.

This is hard to see at the moment cause the cache ttl and our expiry check is using the same TTL. If the TTL cache was 60 mins and our check was 55 minutes, this become evident.

Reply:
As talked about - have removed the additional decorator in favour of simplicity and just have the ttl cache of 55 minutes now